internal/configs/provider.go

```diff
@@ -154,104 +294,108 @@ func (p *Provider) Addr() addrs.LocalProviderConfig {
-// provider address in compact form. The following are examples of traversals
-// that can be successfully parsed as compact relative provider configuration
-// addresses:
-//
-//   - aws
-//   - aws.foo
-//
-// This function will panic if given a relative traversal.
-//
-// If the returned diagnostics contains errors then the result value is invalid
-// and must not be used.
-func ParseProviderConfigCompact(traversal hcl.Traversal) (addrs.LocalProviderConfig, tfdiags.Diagnostics) {
-	var diags tfdiags.Diagnostics
-	ret := addrs.LocalProviderConfig{
-		LocalName: traversal.RootName(),
-	}
+	result := Provider{ProviderCommon: p.ProviderCommon}
+	if p.AliasExpr != nil {
+		if eval != nil {
+			valDiags := eval.DecodeExpression(p.AliasExpr, StaticIdentifier{
+				Module:    eval.call.addr,
+				Subject:   fmt.Sprintf("provider.%s.alias", p.Name),
+				DeclRange: p.AliasExpr.Range(),
+			}, &result.Alias)
+			diags = append(diags, valDiags...)
+		} else {
+			// Test files don't have a static context
+			valDiags := gohcl.DecodeExpression(p.AliasExpr, nil, &result.Alias)
+			diags = append(diags, valDiags...)
+		}
 
-	if len(traversal) < 2 {
-		// Just a type name, then.
-		return ret, diags
+		if diags.HasErrors() {
+			return nil, diags
+		}
+
+		if !hclsyntax.ValidIdentifier(result.Alias) {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid provider configuration alias",
+				Detail:   fmt.Sprintf("Alias %q must be a valid name. %s", result.Alias, badIdentifierDetail),
+				Subject:  p.AliasExpr.Range().Ptr(),
+			})
+		}
 	}
+	return []*Provider{&result}, diags
+}
 
-	aliasStep := traversal[1]
-	switch ts := aliasStep.(type) {
-	case hcl.TraverseAttr:
-		ret.Alias = ts.Name
-		return ret, diags
-	default:
-		diags = diags.Append(&hcl.Diagnostic{
+func (p *ProviderBlock) generateForEachProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if eval == nil {
+		return nil, diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
-			Summary:  "Invalid provider configuration address",
-			Detail:   "The provider type name must either stand alone or be followed by an alias name separated with a dot.",
-			Subject:  aliasStep.SourceRange().Ptr(),
+			Summary:  "Iteration not allowed in test files",
+			Detail:   "for_each was declared as a provider attribute in a test file",
+			Subject:  p.ForEach.Range().Ptr(),
 		})
 	}
 
-	if len(traversal) > 2 {
-		diags = diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid provider configuration address",
-			Detail:   "Extraneous extra operators after provider configuration address.",
-			Subject:  traversal[2:].SourceRange().Ptr(),
-		})
+	forEachRefsFunc := func(refs []*addrs.Reference) (*hcl.EvalContext, tfdiags.Diagnostics) {
+		var diags tfdiags.Diagnostics
+		evalContext, evalDiags := eval.EvalContext(StaticIdentifier{
```

Comment:
This can just be a return, don't need additional vars here

Reply:
it needs them because EvalContext returns `hcl.Diagnostics` but we need `tfdiags.Diagnostics`. So have to do that append magic.

Reply:
Ah, my mistake!